### ISSUE TITLE
NBs for moving to new torch.fft

### ISSUE THREAD

**Issue by galbiati** [pytorch/pytorch#49637]
The new `torch.fft` package is great, but it took me a while to figure out a data format change for going from `torch.irfft` to `torch.fft.irfftn`. `torch.irfft` expects complex inputs to be represented by 2 trailing dimensions, but `torch.fft.irfftn` expects `torch.complex64` (or 128, etc) tensor.

Example:
```python

# spectrum represents complex data with trailing dimension, as for old torch.irfft
# ie, spectrum.size() --> (N, C, D, H, W, 2) for my case

# Old
signal = torch.irfft(spectrum, 3, normalized=True, signal_sizes=(d, h, w))

# Silent bug
signal = torch.fft.irfftn(spectrum, s=(d, h, w), dim=(2, 3, 4), norm='ortho')

# Fixed
spectrum = torch.complex(spectrum[..., 0], spectrum[..., 1])
signal = torch.fft.irfftn(spectrum, s=(d, h, w), dim=(2, 3, 4), norm='ortho')
```

The old and new irfft implementations also don't seem to give exactly the same outputs all the time though. The difference hasn't affected me, but I'm curious where it comes from. Examples:

```python

# Seems close:

torch.manual_seed(12345)
signal = torch.randn(1, 3, 10, 10, 10)
spectrum = torch.rfft(signal, 3, normalized=True)
recovered_old = torch.irfft(spectrum, 3, normalized=True, signal_sizes=(10, 10, 10))
recovered_new = torch.fft.irfftn(torch.complex(spectrum[..., 0], spectrum[..., 1]), dim=(2, 3, 4), s=(10, 10, 10), norm='ortho')

print(recovered_old - recovered_new).abs().max())
# >> Some small number, ~1e-6

# Not so close:
torch.manual_seed(12345)
spectrum = torch.randn(1, 3, 10, 10, 6, 2)    # Maybe this doesn't happen if spectrum is properly scaled?
recovered_old = torch.irfft(spectrum, 3, normalized=True, signal_sizes=(10, 10, 10))
recovered_new = torch.fft.irfftn(torch.complex(spectrum[..., 0], spectrum[..., 1]), dim=(2, 3, 4), s=(10, 10, 10), norm='ortho')

print(recovered_old - recovered_new).abs().max())
# >> Some not so small number, ~1.2
```

Finally, in order to reproduce old behavior closely `norm='ortho'` is the correct value for the new irfftn function. This can be figured out by referring to both functions' documentations, but it'd be nice to have described explicitly in the deprecation notices.

---

Since the old function `torch.irfft` is going to be deprecated in 1.8, it might help some other people to leave some related transitional notes in the documentation on these changes? I couldn't find any notes on this and it was causing errors in my work silently, so it took me some time to figure out what was happening.

**Comment by trip036** [comment]
Definitely amazedSent from my Verizon, Samsung Galaxy smartphone
-------- Original message --------From: gianni <notifications@github.com> Date: 12/18/20  11:52 PM  (GMT-06:00) To: pytorch/pytorch <pytorch@noreply.github.com> Cc: Subscribed <subscribed@noreply.github.com> Subject: [pytorch/pytorch] NBs for moving to new torch.fft (#49637) 
The new torch.fft package is great, but it took me a while to figure out a data format change for going from torch.irfft to torch.fft.irfftn. torch.irfft expects complex inputs to be represented by 2 trailing dimensions, but torch.fft.irfftn expects torch.complex64 (or 128, etc) tensor.
Example:
# spectrum represents complex data with trailing dimension, as for old torch.irfft
# ie, spectrum.size() --> (N, C, D, H, W, 2) for my case

# Old
signal = torch.irfft(spectrum, 3, normalized=True, signal_sizes=(d, h, w))

# Silent bug
signal = torch.fft.irfftn(spectrum, s=(d, h, w), dim=(2, 3, 4), norm='ortho')

# Fixed
spectrum = torch.complex(spectrum[..., 0], spectrum[..., 1])
signal = torch.fft.irfftn(spectrum, s=(d, h, w), dim=(2, 3, 4), norm='ortho')
The old and new irfft implementations also don't seem to give exactly the same outputs all the time though. The difference hasn't affected me, but I'm curious where it comes from. Examples:
# Seems close:

torch.manual_seed(12345)
signal = torch.randn(1, 3, 10, 10, 10)
spectrum = torch.rfft(signal, 3, normalized=True)
recovered_old = torch.irfft(spectrum, 3, normalized=True, signal_sizes=(10, 10, 10))
recovered_new = torch.fft.irfftn(torch.complex(spectrum[..., 0], spectrum[..., 1]), dim=(2, 3, 4), s=(10, 10, 10), norm='ortho')

print(recovered_old - recovered_new).abs().max())
# >> Some small number, ~1e-6

# Not so close:
torch.manual_seed(12345)
spectrum = torch.randn(1, 3, 10, 10, 6, 2)    # Maybe this doesn't happen if spectrum is properly scaled?
recovered_old = torch.irfft(spectrum, 3, normalized=True, signal_sizes=(10, 10, 10))
recovered_new = torch.fft.irfftn(torch.complex(spectrum[..., 0], spectrum[..., 1]), dim=(2, 3, 4), s=(10, 10, 10), norm='ortho')

print(recovered_old - recovered_new).abs().max())
# >> Some not so small number, ~1.2
Finally, in order to reproduce old behavior closely norm='ortho' is the correct value for the new irfftn function. This can be figured out by referring to both functions' documentations, but it'd be nice to have described explicitly in the deprecation notices.

Since the old function torch.irfft is going to be deprecated in 1.8, it might help some other people to leave some related transitional notes in the documentation on these changes? I couldn't find any notes on this and it was causing errors in my work silently, so it took me some time to figure out what was happening.

**Comment by mruberry** [comment]
Hey @galbiati, thanks for filing this issue, let's go through it together.

> The new `torch.fft` package is great

Thank you for saying this. It was developed by @peterbell10. 

> but it took me a while to figure out a data format change for going from `torch.irfft` to `torch.fft.irfftn`. `torch.irfft` expects complex inputs to be represented by 2 trailing dimensions, but `torch.fft.irfftn` expects `torch.complex64` (or 128, etc) tensor.

Correct. Sorry this was confusing. Maybe we should think about making this more prominent in the new torch.fft documentation. We could write a paragraph or two about the entire module, for example. 

> The old and new irfft implementations also don't seem to give exactly the same outputs all the time though.

This is interesting. 

> 
> ```python
> torch.manual_seed(12345)
> signal = torch.randn(1, 3, 10, 10, 10)
> spectrum = torch.rfft(signal, 3, normalized=True)
> recovered_old = torch.irfft(spectrum, 3, normalized=True, signal_sizes=(10, 10, 10))
> recovered_new = torch.fft.irfftn(torch.complex(spectrum[..., 0], spectrum[..., 1]), dim=(2, 3, 4), s=(10, 10, 10), norm='ortho')
> 
> print(recovered_old - recovered_new).abs().max())
> # >> Some small number, ~1e-6

I can confirm that torch.irfftn, torch.fft.irfftn, and np.fft.irfftn all produce different values. Taking the sum of the errors gives an error of ~4e-4. 
 
> ```python
> # Not so close:
> torch.manual_seed(12345)
> spectrum = torch.randn(1, 3, 10, 10, 6, 2)    # Maybe this doesn't happen if spectrum is properly scaled?
> recovered_old = torch.irfft(spectrum, 3, normalized=True, signal_sizes=(10, 10, 10))
> recovered_new = torch.fft.irfftn(torch.complex(spectrum[..., 0], spectrum[..., 1]), dim=(2, 3, 4), s=(10, 10, 10), norm='ortho')
> 
> print(recovered_old - recovered_new).abs().max())
> # >> Some not so small number, ~1.2
> ```

The sum of errors for this case is ~948 for me, which is significant. What's interesting is that np.fft.irfftn produces results similar to torch.fft.irfftn. This suggests:

- torch.irfftn was incorrect
- torch.fft.irfftn and np.fft.irfftn are incorrect
- there is some issue with the snippet's translation of torch.irfft to torch.fft.irfftn 

I'm guessing the problem here is the third option, and that there is a subtle issue with the handling of signal dimensions. For example, in the second example "spectrum" is not constructed from a call to torch.rfft. What do you think, @peterbell10? 
 
> Finally, in order to reproduce old behavior closely `norm='ortho'` is the correct value for the new irfftn function. This can be figured out by referring to both functions' documentations, but it'd be nice to have described explicitly in the deprecation notices.

This is the case when torch.rfft and torch.irfft were called with `normalized=True` but the default is `normalized=False`. A mapping from the optional settings to the new ones would still be nice.

> Since the old function `torch.irfft` is going to be deprecated in 1.8, it might help some other people to leave some related transitional notes in the documentation on these changes? I couldn't find any notes on this and it was causing errors in my work silently, so it took me some time to figure out what was happening.

That's true. We have this wiki article: https://github.com/pytorch/pytorch/wiki/The-torch.fft-module-in-PyTorch-1.7 but it doesn't get into the more complicated cases of optional values, like we're looking at with these snippets.

**Comment by peterbell10** [comment]
Yes, `torch.irfft` and `torch.fft.irfftn` behave slightly differently, but only on invalid input. If you think about the `irfftn` to be an `ifftn` followed by an `irfft` then on that last complex-to-real transform the first and possibly last values in the transformed dimension should have 0 imaginary component. This is part of the hermitian-symmetry property that must be true for the output is real. By randomly generating data, this property probably won't be satisfied.

`numpy.fft` and `torch.fft` give the result you get by zeroing out the imaginary components of these values. The old functions give some other result but I'm not sure if there's any reasoning behind it.

**Comment by peterbell10** [comment]
Here is an example of not using `rfftn` but still ensuring the data satisfies the Hermititan property:
```python
In [1]: import torch.fft
   ...: torch.manual_seed(12345)
   ...: spectrum = torch.randn(1, 3, 10, 10, 6, dtype=torch.cfloat)    # Maybe this doesn't happen if spectrum is properly scaled?
   ...: spectrum[..., 0].imag.fill_(0)
   ...: spectrum[..., -1].imag.fill_(0)
   ...: spectrum = torch.view_as_real(torch.fft.fftn(spectrum, dim=(2, 3)))
   ...: recovered_old = torch.irfft(spectrum, 3, normalized=True, signal_sizes=(10, 10, 10))
   ...: recovered_new = torch.fft.irfftn(torch.complex(spectrum[..., 0], spectrum[..., 1]), dim=(2, 3, 4), s=(10, 10, 10), norm='ortho')
   ...: (recovered_old - recovered_new).abs().max()
Out[1]: tensor(7.6294e-06)
```

**Comment by galbiati** [comment]
@mruberry 
> That's true. We have this wiki article: https://github.com/pytorch/pytorch/wiki/The-torch.fft-module-in-PyTorch-1.7 but it doesn't get into the more complicated cases of optional values, like we're looking at with these snippets.

Ah, thanks for this link! If I had found this before, it would have been perfect :) Maybe the link could be included in the 1.8 release notes?

@peterbell10
> By randomly generating data, this property probably won't be satisfied.

This makes perfect sense. In my actual application I'm only generating the real component randomly; didn't think it through when generating these examples. Thanks for the clarification!

**Comment by mruberry** [comment]
Leaving this open for the moment while we think how best to document torch.fft in 1.8. Glad to see this particular issue was resolved, however.

**Comment by trip036** [comment]
https://www.github.com/pytorch/rfcs/tree/master/LICENSESent from my Verizon, Samsung Galaxy smartphone

**Comment by vadimkantorov** [comment]
Is it okay that old rfft supported onesided argument, and the new one only implements onesided=True. I propose to either support this flag, or support it and print a porting guide if it's used. May be easier to just support it @mruberry 

A real-life usecase needing rfft(onesided = False): abusing rfft for computing sin/cos values in order to simulate stft with Conv1d: https://github.com/vadimkantorov/convasr/blob/f241919/models.py#L521

**Comment by mruberry** [comment]
@peterbell10, what do you think about updating fft.rfft() with a kwarg?

**Comment by peterbell10** [comment]
`rfft(x, onesided=False)` is just `fft(x)`.

**Comment by eric-vader** [comment]
Hi sorry for reopening this again... But I am porting over some old code and I have a problem replicating the same results with the new API; I have read the above discussion and tried to find the corresponding new API and here is what I got:
```
# old
print(t.size()) # torch.Size([3, 32, 32, 2])
r = torch.irfft(t, 3, normalized=True, onesided=False)
# new
_t = torch.view_as_complex(t)
_new = torch.fft.ifftn(_t, dim=(0,1,2), s=(3, 32, 32), norm='ortho')
```
However, the results are incorrect (old and new are different).
Any help will be greatly appreciated!

**Comment by peterbell10** [comment]
@eric-vader have you tried `irfftn`?

**Comment by eric-vader** [comment]
Yeap

for ifftn:
`print((res[i] - _new).abs().max())`
tensor(0.0133)

for irfftn:
`print((res[i] - _new).abs().max())`
tensor(0.0036)

Doesnt seem to be correct; am I calling the new API correct, equivalently?

**Comment by peterbell10** [comment]
Yes, I think the `irfftn` version should be equivalent. Do you have a complete script that reproduces the issue?

**Comment by eric-vader** [comment]
Hi, thanks for the reply, and sorry for the late reply as I am traveling!

I am trying to port the code in the repo [https://github.com/satyanshukla/bayes_attack](https://github.com/satyanshukla/bayes_attack)

specifically in utils.py:
* [Line 61](https://github.com/satyanshukla/bayes_attack/blob/f2fdd378cadc23bca7ee45c5651f3cdc5b428c1b/utils.py#L61)
* [Line 80](https://github.com/satyanshukla/bayes_attack/blob/f2fdd378cadc23bca7ee45c5651f3cdc5b428c1b/utils.py#L80)
